replace gulp with webpack, update dependencies, add standalone template

andreasgerstmayr opened the pull request.

spiermar commented:
@andreasgerstmayr do you have an example of the template file being used? With the stacks inlined?

andreasgerstmayr commented:
Yep: https://gist.github.com/andreasgerstmayr/e7f76b6afa8f2ef081ffa96f952ff428

spiermar commented on webpack.config.js:
```diff
@@ -0,0 +1,109 @@
+module.exports = [{
+  context: path.join(__dirname, 'src'),
+  entry: {
+    './d3-flamegraph': './flamegraph.js',
```
Noticed the output of webpack is significantly larger than rollup. ~80%. I was expecting it to be larger, because of the extra webpack magic, but not that much. Wondering if it's importing more files or duplicating imports.

andreasgerstmayr replied:
`d3-flamegraph.min.js` by rollup is 59 KB, compiled with webpack it is 78 KB (+ 32%). The development version is quite big, but I wouldn't worry about that. The colormapper is bigger (+1 KB), that's because of the webpack header.

About the 59 vs 78 KB: I think rollup is more aggressively deleting dead code. For example I can see the code `schedule = schedules[i]).state > SCHEDULED` of `node_modules/d3-transition/src/active.js` in the webpack version, but not the rollup version. Which looks like unused code.

spiermar commented on src/template.css:
```diff
@@ -0,0 +1,84 @@
+/*
```
Would avoid adding licensed code here.

andreasgerstmayr replied:
What do you suggest? Redesign the template?
It's the few selectors from bootstrap which we actually use in the template (before I included the full bootstrap CSS, but then the empty template was 238 KB)

spiermar replied:
Assuming you will update the PR with a template similar to the one I've sent you.

andreasgerstmayr replied:
done

spiermar commented on webpack.config.js:
```diff
@@ -0,0 +1,109 @@
+const { CleanWebpackPlugin } = require('clean-webpack-plugin');
+const CopyWebpackPlugin = require('copy-webpack-plugin');
+const HtmlWebpackPlugin = require('html-webpack-plugin');
+const HtmlWebpackInlineSourcePlugin = require('html-webpack-inline-source-plugin');
```
Looks abandoned.

andreasgerstmayr replied:
Ok, I replaced `html-webpack-inline-source-plugin` with `script-ext-html-webpack-plugin`.

spiermar commented on src/template.js:
```diff
@@ -0,0 +1,50 @@
+
+class FlameGraphUI {
+    constructor (stacks, options) {
+        this.stacks = stacks || { name: 'placeholder', value: 0, children: [] }
```
Might want to leave undefined and fail if stack is not defined.

andreasgerstmayr replied:
Fail with a JS error? I put the placeholder here so for ppl who don't know about the JS console etc. who open the empty template by mistake see something. I've updated the text now to `Please replace /** @flamegraph_params **/ with flame graph stacks in the template.`

spiermar replied:
Thinking more of an error message saying stacks were not loaded.

andreasgerstmayr replied:
done

andreasgerstmayr commented on webpack.config.js:
```diff
@@ -0,0 +1,109 @@
+      template: 'template.html',
+      filename: 'template.html',
+      inject: 'head',
+      inlineSource: '.(js|css)$'
```
just checked the compiled template file - it's not included (webpack only includes the files which are imported from the `template.js`)

andreasgerstmayr commented:
Thanks for the review! I updated the PR with a new commit.

There are only two remaining issues
* the bigger size (19 KB) - maybe there is some plugin parameter to more aggressively delete unused code? I'll see what I can do.
* re bootstrap: I've already removed the full bootstrap CSS, that's why I extracted parts of it to `template.css`, not sure how I should proceed here (redesign the template from scratch?)

spiermar commented:
I'm Ok with the size difference for now, as long as we know where it is coming from. We can work on optimizing it later.

andreasgerstmayr commented:
I've update the template now and rebased it on latest master.

andreasgerstmayr commented:
I've added a `template_version` meta tag with the content of the current version (of `package.json`) and renamed the template variable to `@flamegraph_json`.

andreasgerstmayr commented:
also added a loading text (loading ~300MB flame graph files takes a while)
